src/chapter_3/CPIs.md

```diff
@@ -0,0 +1,306 @@
+}
+```
+
+Also add the line `puppet_master = "HmbTLCmaGvZhKnn1Zfa1JVnp7vkMV4DYVxPLWBVoN65L"` in the `[programs.localnet]` section of your `Anchor.toml`. Finally, import the puppet program into the puppet-master program by adding the following line to the `[dependencies]` section of the `Cargo.toml` file inside the `puppet-master` program folder:
+```toml
+puppet = { path = "../puppet", features = ["cpi"]}
+```
+
+The `features = ["cpi"]` is used so we can not only use puppet's types but also its instruction builders and cpi functions. Without those, we would have to use low level solana syscalls. Fortunately, anchor provides abstractions on top of those. By enabling the `cpi` feature, the puppet-master program gets access to the `puppet::cpi` module. Anchor generates this module automatically and it contains tailor-made instructions builders and cpi helpers for the program.
+
+In the case of the puppet program, the puppet-master uses the `SetData` instruction builder struct provided by the `puppet::cpi::accounts` module to submit the accounts the `SetData` instruction of the puppet program expects. Then, the puppet-master creates a new cpi context and passes it to the `puppet::cpi::set_data` cpi function. This function has the exact same function as the `set_data` function in the puppet program with the exception that it expects a `CpiContext` instead of a `Context`.
+
+We can verify that everything worked as expects by replacing the contents of the `puppet.ts` file with:
+```ts
+import * as anchor from '@project-serum/anchor';
+import { web3 } from '@project-serum/anchor/';
+import { Program } from '@project-serum/anchor';
+import { Puppet } from '../target/types/puppet';
+import { PuppetMaster } from '../target/types/puppet_master';
+import { expect } from 'chai';
+
+describe('puppet', () => {
+  anchor.setProvider(anchor.Provider.env());
+
+  const puppetProgram = anchor.workspace.Puppet as Program<Puppet>;
+  const puppetMasterProgram = anchor.workspace.PuppetMaster as Program<PuppetMaster>;
+
+  const puppetKeypair = web3.Keypair.generate();
+
+  it('Does CPI!', async () => {
+    await puppetProgram.rpc.initialize({
+      accounts: {
+        puppet: puppetKeypair.publicKey,
+        user: anchor.getProvider().wallet.publicKey,
+        systemProgram: web3.SystemProgram.programId
+      },
+      signers: [puppetKeypair]
+    });
+
+    await puppetMasterProgram.rpc.pullStrings(new anchor.BN(42),{
+      accounts: {
+        puppetProgram: puppetProgram.programId,
+        puppet: puppetKeypair.publicKey
+      }
+    })
+
+    expect((await puppetProgram.account.data
+      .fetch(puppetKeypair.publicKey)).data.toNumber()).to.equal(42);
+  });
+});
+```
+
+## Privilege Extension
+
+CPIs extend the privileges of the caller to the callee. The puppet account was passed as a mutable account to the puppet-master but it was still mutable in the puppet as well (otherwise the `expect` in the test would've failed). The same applies to signatures.
```

Comment:
Maybe we should add some type of additional security comment here saying that the developer should be thoughtful / careful when using CPI to an unknown program.

Reply:
added this:

> Privilege extension is convenient but also dangerous. If a CPI is unintentionally made to a malicious program,
> this program has the same privileges as the caller.
> Anchor protects you from CPIs to malicious programs with two measures.
> First, the `Program<'info, T>` type checks that the given account is the expected program `T`.
> Should you ever forget to use the `Program` type, the automatically generated cpi function
> (in the previous example this was `puppet::cpi::set_data`)
> also checks that the `cpi_program` argument equals the expected program.